### PR TITLE
Tested synchronization optimization 

### PR DESCRIPTION
**Pull request by kubazak**
Optimalization of import from external source 

ExtSource*:
 - Methods getGroupSubjects now return string with status. It fills subjects through given parameter now.
 - Private methods changed to support change mentioned above.

Methods categorizeMembersForLightweightSynchronization and categorizeMembersForSynchronization were changed.
Now it uses map where key is login from needed external source and value is RichMember attached to this login.
Thanks to this is checking if user is already in group faster.

Methods to getCandidate were refactored. Now they count hash code of subject obtained from external source.
Hashcodes of subjects are stored in member-group attribute. If it has same value in the next synchronization cycle,
we don't need to update this member's attributes.

Communication with external sources SQL, SQLComplex and LDAP was changed.
Now they can send us only modified members of group, when it is possible.

UserDuplicateException was created. It is thrown when candidate,
which should be added to group, represents more than one User.

Problem:
When Admin Perun change extSource file Members from previous extSoure stayed
in the database.

Fix:
I added the condition if (member doesn't have same extSource as actual
extSource, then delete it.) to the method getMapOfLoginsAndRichMembers.

Delete unnecessary code

Problem:
When is testing synchronization where members are inserted to the group
and users are in the group so synchronization is much long
(3x longer than actual master).

Fix:
I deleted code where repeat created attribute hashCode which is created
in method updateExistingMembersWhileSynchronization.

### CODE REVIEW

**Comment by stavamichal** [comment]
The main message shouldn't be just about fixes but about the whole concept (from the first commit message).

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 1]
```diff
@@ -2,6 +2,39 @@
...
+import org.springframework.jdbc.core.RowMapper;
+
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+
+
+import org.springframework.jdbc.support.lob.DefaultLobHandler;
+import org.springframework.jdbc.support.lob.LobHandler;
+
+import cz.metacentrum.perun.core.api.BeansUtils;
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: Why did you add these imports?

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java [hunk 2]
```diff
@@ -69,16 +76,17 @@ public static PerunBlImpl setPerunBlImpl(PerunBlImpl perun) {
             //Replace '?' by searchString
             query = query.replaceAll("\\?", searchString);
 
-            //Get csv file 
-            prepareEnvironment();
+            //Get csv file
+            prepareFile();
 
-            return csvParsing(query, maxResults);
+			List<Map<String, String>> subjects = new ArrayList<>();
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java]: Wrong indentation.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java [hunk 3]
```diff
@@ -113,71 +122,111 @@ public static PerunBlImpl setPerunBlImpl(PerunBlImpl perun) {
...
+     * @param group Map of attributes used for quering the external source
+     * @param status Status about synchronization (lightweight, full). Not used in CSV external source.
+     * @param subjects list of maps, which contains attr_name-&gt;attr_value, e.g. firstName-&gt;Michal
+     * @return String containing status about data received from external source (e.g. data are sorted,
+     * @throws InternalErrorException
+     * @throws ExtSourceUnsupportedOperationException
+     */
     @Override
-    public List<Map<String, String>> getGroupSubjects(Map<String, String> attributes) throws InternalErrorException, ExtSourceUnsupportedOperationException {
+    public String getGroupSubjects(PerunSession sess, Group group, String status, List<Map<String, String>> subjects) throws InternalErrorException, ExtSourceUnsupportedOperationException {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java]: This is not correct. Method says "Get subjects of group" in javadoc. But instead of list of subjects, it returns some sort of internal status.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java [hunk 4]
```diff
@@ -113,71 +122,111 @@ public static PerunBlImpl setPerunBlImpl(PerunBlImpl perun) {
...
                 throw new InternalErrorException("Attribute " + GroupsManager.GROUPMEMBERSQUERY_ATTRNAME + " can't be null.");
             }
 
-            //Get csv file
-            prepareEnvironment();
+            // Get CSV file
+            prepareFile();
 
-            return csvParsing(queryForGroup, 0);
+            csvParsing(queryForGroup, 0, subjects);
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java]: Again method with side effects, which I prefer to prevent if possible.

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/GroupsManager.java [hunk 5]
```diff
@@ -16,28 +44,44 @@
...
  * @author  Slavek Licehammer
+ * @author  Jan Zverina
  * @see Perun
  */
 public interface GroupsManager {
 
 	// Attributes related to the external groups
 	// Contains query need to get the group members
-	public static final String GROUPMEMBERSQUERY_ATTRNAME = AttributesManager.NS_GROUP_ATTR_DEF + ":groupMembersQuery";
+	String GROUPMEMBERSQUERY_ATTRNAME = AttributesManager.NS_GROUP_ATTR_DEF + ":groupMembersQuery";
```
[the remark this answers is not in the document]
Comment (kubazak) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/GroupsManager.java]: Done

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/MembersManager.java [hunk 6]
```diff
@@ -1151,7 +1180,7 @@
 	 * @throws UserNotInRoleException if sponsor hasn't right role in the same vo
 	 * @throws PrivilegeException if not PerunAdmin
 	 */
-	RichMember setSponsorshipForMember(PerunSession session, Member sponsoredMember, User sponsor) throws InternalErrorException, MemberNotExistsException, AlreadySponsoredMemberException, UserNotInRoleException, PrivilegeException;
+	RichMember setSponsorshipForMember(PerunSession session, Member sponsoredMember, User sponsor) throws InternalErrorException, MemberNotExistsException, UserNotInRoleException, AlreadySponsoredMemberException, PrivilegeException;
```
[the remark this answers is not in the document]
Comment (kubazak) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/MembersManager.java]: Done

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ExtSourcesManagerBlImpl.java [hunk 7]
```diff
@@ -229,63 +231,88 @@ public Candidate getCandidate(PerunSession sess, ExtSource source, String login)
...
+	 *
+	 * @param sess Perun session
+	 * @param subject Subject to gain attributes from
+	 * @param login Login of candidate
+	 * @param attributes Attributes of candidate, which will be filled
+	 * @param additionalUserExtSources Additional user external sources of candidate, which will be filled
+	 * @throws InternalErrorException
+	 */
+	private void filterAttributes(PerunSession sess, Map<String, String> subject, String login,
+												 Map<String, String> attributes, List<UserExtSource> additionalUserExtSources) throws InternalErrorException {
```
[the remark this answers is not in the document]
Comment (kubazak) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ExtSourcesManagerBlImpl.java]: Done

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java (hunk 1 above)
Comment (kubazak) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: It was created when merging. It was deleted.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java (hunk 2 above)
Comment (kubazak) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceCSV.java]: Done.

**Comment by balcirakpeter** [comment]
This pull request will be split into more smaller pull requests. We can close this one when the smaller pull requests will be merged.
